numbast/src/numbast/static/struct.py

```diff
@@ -0,0 +1,691 @@
+
+        Parameter
+        ---------
+        signature_strs: list[str]
+            A list of `numba.signature` strings containing all overloads of
+            the constructors.
+        """
+
+        self.Imports.add("from numba.cuda.cudadecl import register")
+        self.Imports.add("from numba.cuda.cudadecl import register_global")
+        self.Imports.add("from numba.core.typing.templates import ConcreteTemplate")
+        self.Imports.add("from numba.types import Function")
+
+        signatures_str = ", ".join(signature_strs)
+
+        self._struct_ctor_typing_rendered = (
+            self.struct_ctor_template_typing_template.format(
+                struct_name=self._struct_name, signatures=signatures_str
+            )
+        )
+
+    def _render(self):
+        """Render all struct constructors."""
+
+        signatures: list[str] = []
+        for ctor_decl in self._ctor_decls:
+            renderer = StaticStructCtorRenderer(
+                struct_name=self._struct_name,
+                struct_type_class=self._struct_type_class,
+                struct_type_name=self._struct_type_name,
+                ctor_decl=ctor_decl,
+            )
+            renderer._render()
+
+            self._python_rendered += renderer._python_rendered
+            self._c_rendered += renderer._c_rendered
+
+            signatures.append(renderer.signature_str)
+
+        self._render_typing(signatures)
+
+        self._python_rendered += self._struct_ctor_typing_rendered
+
+    @property
+    def python_rendered(self) -> str:
+        """The python script that contains the bindings to all constructors."""
+        return self._python_rendered
+
+    @property
+    def c_rendered(self) -> str:
+        """The C program that contains the shim functions to all constructors."""
+        return self._c_rendered
+
+
+class StaticStructRenderer(BaseRenderer):
+    """Renderer that renders bindings to a single CUDA C++ struct.
+
+    Parameters
+    ----------
+    decl: Struct
+        Declaration of the struct.
+    parent_type: type
+        The parent Numba type of the new struct type created.
+    data_model: type
+        The data model of the new struct type.
+    header_path: os.PathLike
+        Path to the header that contains the declaration of the struct.
+    aliases: list[str], optional
+        TODO: If the struct has other aliases, specify them here. Numbast creates
+        aliased objects that references the original python API object.
+    """
+
+    typing_template = """
+# Typing for {struct_name}
+class {struct_name}_type_class({parent_type}):
+    def __init__(self):
+        super().__init__(name="{struct_name}")
+        self.alignof_ = {struct_alignof}
+        self.bitwidth = {struct_sizeof} * 8
+
+{struct_name}_type = {struct_name}_type_class()
+"""
+
+    python_api_template = """
+# Make Python API for struct
+{struct_name} = type("{struct_name}", (), {{"_nbtype": {struct_name}_type}})
```

Comment:
Is this expected to be used for something? I'm not sure if it's of any use or connected to your intentions here, but you can provide a `_numba_type_` property that Numba will use to get the type instead of adding typing for it - for example: https://github.com/numba/numba/blob/main/numba/cuda/cudadrv/devicearray.py#L169

Reply:
Now I think about it / recall a bit more, I think the `_numba_type_` property is only used when you're passing something in from the Python side as an argument so it probably doesn't help much here (at least until/if we want to start supporting passing in Python objects that are instances of generated structs)

Reply:
Yes, I believe `_nbtype` was used mainly in the tests and the pytorch bindings to provide a handy access to the numba type. It is not meant to be used by users though.

https://github.com/NVIDIA/numbast/blob/ddb57d14c46def9050cc3d9f763117203db9cdb9/numbast/tests/test_struct.py#L27

https://github.com/NVIDIA/numbast/blob/7f36a8c628a06e6cc36d64e7b0c43b99c4dc4f2a/numbast_extensions/numbast_extensions/bf16/torch_interop.py#L104

Reply:
Ah, I see - I didn't notice the use of it at the time because I was focusing only on the code I generated when experimenting. Thanks for the clarification.